packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m

```diff
@@ -21,26 +22,29 @@ @interface FVPFrameUpdater : NSObject
 @property(nonatomic, weak, readonly) NSObject<FlutterTextureRegistry> *registry;
 // The output that this updater is managing.
 @property(nonatomic, weak) AVPlayerItemVideoOutput *videoOutput;
-// The last time that has been validated as avaliable according to hasNewPixelBufferForItemTime:.
-@property(nonatomic, assign) CMTime lastKnownAvailableTime;
+// The display link that drives frameUpdater.
+@property(nonatomic) FVPDisplayLink *displayLink;
+// The time interval between screen refresh updates.
+@property(nonatomic) _Atomic CFTimeInterval duration;
 @end
 
 @implementation FVPFrameUpdater
 - (FVPFrameUpdater *)initWithRegistry:(NSObject<FlutterTextureRegistry> *)registry {
   NSAssert(self, @"super init cannot be nil");
   if (self == nil) return nil;
   _registry = registry;
-  _lastKnownAvailableTime = kCMTimeInvalid;
   return self;
 }
 
 - (void)displayLinkFired {
-  // Only report a new frame if one is actually available.
-  CMTime outputItemTime = [self.videoOutput itemTimeForHostTime:CACurrentMediaTime()];
-  if ([self.videoOutput hasNewPixelBufferForItemTime:outputItemTime]) {
-    _lastKnownAvailableTime = outputItemTime;
-    [_registry textureFrameAvailable:_textureId];
-  }
+  // Display link duration is in an undefined state until displayLinkFired is called at least once
+  // so it should not be used directly.
+  atomic_store_explicit(&_duration, _displayLink.duration, memory_order_relaxed);
+  [_registry textureFrameAvailable:_textureId];
```

Comment:
In such a case I could just use `targetTimestamp` from the display link. This would be that [previous implementation](https://github.com/flutter/packages/pull/7466/commits/ad0ba7d5d76624e685dbe73980ef452b3e9f364f) just with added `targetTimestamp` and I had it ready at some time for both ios and macos and although it was better with `targetTimestamp` instead of `CACurrentMediaTime` it still had that inherent problem with race which this new version solves.

Engine calls `copyPixelBuffer` only once per frame regardless of how many times `textureFrameAvailable` was called (its effect took place) and that is actually the problem. If it was possible to use `textureFrameAvailable` to enqueue multiple (at least 2) calls to `copyPixelBuffer` (provided that they will not be called at the same screen update) then that previous implementation could be fine.